summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala

```diff
@@ -1,102 +1,238 @@
-    }
-
-  private def registerEdges(topologyWithNodes: Topology): Topology =
-    stormDag.nodes.foldLeft(topologyWithNodes) {
-      case (topology, node: SummerNode[Storm]) =>
-        val edgeType = EdgeType.AggregatedKeyValues[Any, Any](getSummerKeyValueShards(node))
-        registerIncomingEdges(topology, node, edgeType)
-      case (topology, node: FlatMapNode[Storm]) =>
-        registerIncomingEdges(topology, node, flatMapEdgeType(node))
-      case (topology, node: SourceNode[Storm]) => topology
-        // ignore, no incoming edges into sources
+  def build: StormTopology = {
+    val topology = stormDag.nodes.foldLeft(Topology.empty) {
+      case (currentTopology, node: SummerNode[Storm]) =>
+        register(currentTopology, node, SummerBoltProvider(this, node))
+      case (currentTopology, node: FlatMapNode[Storm]) =>
+        register(currentTopology, node, FlatMapBoltProvider(this, node))
+      case (currentTopology, node: SourceNode[Storm]) =>
+        register(currentTopology, node, SpoutProvider(this, node))
     }
+    topology.build(jobId)
+  }
 
-  private def flatMapEdgeType(node: FlatMapNode[Storm]): EdgeType[_] = {
-    val usePreferLocalDependency = getOrElse(node, DEFAULT_FM_PREFER_LOCAL_DEPENDENCY)
-    logger.info(s"[${getNodeName(node)}] usePreferLocalDependency: ${usePreferLocalDependency.get}")
-    if (usePreferLocalDependency.get) {
-      EdgeType.itemWithLocalOrShuffleGrouping[Any]
+  private def register(topology: Topology, node: StormNode, provider: ComponentProvider): Topology =
+    if (!shouldEmitKeyValues(node)) {
+      registerItemComponent(topology, node, provider)
     } else {
-      EdgeType.itemWithShuffleGrouping[Any]
+      registerKeyValueComponent(topology, node, provider)
     }
+
+  private def registerItemComponent[T](
+    topology: Topology,
+    node: StormNode,
+    provider: ComponentProvider
+  ): Topology = {
+    val component = provider.createSingle(identity[Item[T]])
+    val (componentId, topologyWithComponent) = topology.withComponent(getNodeName(node), component)
+    registerItemEdges[T](topologyWithComponent, node, componentId)
   }
 
-  private def registerIncomingEdges(topology: Topology, node: StormNode, edgeType: EdgeType[_]): Topology = {
-    val nodeId: ReceivingId[Any] = getId(node).asInstanceOf[Topology.ReceivingId[Any]]
-    stormDag.dependenciesOf(node).foldLeft(topology) { (current, upstreamNode) =>
-      current.withEdge(Topology.Edge[Any](
-        getId(upstreamNode),
-        edgeType.asInstanceOf[EdgeType[Any]],
-        nodeId))
-    }
+  private def registerKeyValueComponent[K, V](
+    topology: Topology,
+    node: StormNode,
+    provider: ComponentProvider
+  ): Topology = outgoingSummersProps(node) match {
+    case Some(props) if props.allSummers =>
+      // Use Aggregated if possible.
+      provider.createAggregated[K, V](props.batcher, props.shards, props.semigroup
+        .asInstanceOf[Semigroup[V]]) match {
```

Comment:
This asInstanceOf hanging on second line is a bit weird. Can we move this with what it's acting on? If needed we can add a temporary variable.